[CORE-8797] Improve parsing of POST /subject/{subject}/version object

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Fixes: #24782

TODO:

- [x] write tests

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v24.3.x
- [X] v24.2.x
- [X] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

### Bug Fixes

* Fixes integer overflow issues when given a schema via the `POST /subject/{subject}/version` where version was > INT_MAX or a negative value was provided.

michael-redpanda committed:
sr: Updated handling of integer values

For POST /subject/{subject}/version, updating the JSON object parsing to
properly handle situations where a negative value is provided or a value
greater than INT_MAX is provided.  If a negative value is provided,
convert it to the appropriate "invalid" type so it can be automatically
determined.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
sr: Added error for version exhaustion

Version exhaustion can occur when attempting to insert schemas using
version '0' and the maximum version for the subject is INT_MAX.  This
differs slightly from upstream implementation wherein the version is
reused and previous instance of the version is overwritten.

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b81110ee-df66-4947-9956-bc74a4b6630c-->
<!--TEST RESULTS-->
<!--8e14abd5-72d1-49a1-a0bc-8e3337ba8ae5-->
<!--TEST RESULTS-->
<!--7d071dfb-72bb-4c40-8a64-f45ba2289fac-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60928">build#60928</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_single_partition_leadership_transfer.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60928#01947653-7c1f-43db-903c-c88b51a94b73 | FLAKY | 1/2 |
| rptest.tests.consumer_group_balancing_test.ConsumerGroupBalancingTest.test_coordinator_nodes_balance | ducktape | https://buildkite.com/redpanda/redpanda/builds/60928#01947653-7c1c-44e7-9e32-6c9035603d34 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61013">build#61013</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61013#01948b3c-cc4d-4882-b9de-59eb3c224160 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61120">build#61120</a></summary>


| test_id                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61120#019494c4-cafd-4c80-bdfd-c81b004823cb | FLAKY         | 1/2      |
</details>
<!--/body-->

pgellert commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -466,6 +466,16 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
+        unparsed.version = invalid_schema_version;
+    }
+
+    // Upstream permits IDs of 0 on 'import'
```
I think it's not just in 'import' mode. I think it's the version, not the id, that is only allowed to be 0 on import mode. And the id just has to be unique but can be 0. Eg. this works for me out of the box (ie. not in import mode):
```
% curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"id": 0, "version": 1, "schema": "{\"name\": \"Foo\", \"type\": \"record\", \"fields\": [{\"name\": \"data\", \"type\": \"float\"}]}"}' http://localhost:8081/subjects/baz/versions
```

michael-redpanda replied:
> Eg. this works for me out of the box (ie. not in import mode)

I just tried this and it got rejected:

```
{"error_code":42205,"message":"Subject baz is not in import mode"}
```

michael-redpanda replied:
This is using version 7.8.0 of SR

pgellert replied:
Sorry, my mistake here. That's what I'm seeing as well. I think I must have had import mode on and thought it wasn't import mode when testing this.

pgellert commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -466,6 +466,16 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
     auto unparsed = co_await ppj::rjson_parse(
       std::move(rq.req), post_subject_versions_request_handler<>{sub});
 
+    // If version or ID is not a positive integer, then set version to invalid_schema_version
+    if (unparsed.version.has_value() && unparsed.version.value() < 1) {
+        unparsed.version = invalid_schema_version;
+    }
```
Is it correct to reject `version=0` here? In import mode, the reference implementation accepts `id=0; version=0` and treats that as `id=0; version=${next_available}`. 
```
% curl -X PUT -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"mode": "IMPORT"}' http://localhost:8081/mode
{"mode":"IMPORT"}

% curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"id": 0, "version": 0, "schema": "{\"name\": \"Foo\", \"type\": \"record\", \"fields\": [{\"name\": \"data\", \"type\": \"float\"}]}"}' http://localhost:8081/subjects/baz/versions
{"id":0}

% curl http://localhost:8081/subjects/baz/versions/1
{"subject":"baz","version":1,"id":0,"schema":"{\"type\":\"record\",\"name\":\"Foo\",\"fields\":[{\"name\":\"data\",\"type\":\"float\"}]}"}

% curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"id": 1, "version": 0, "schema": "{\"name\": \"Foo2\", \"type\": \"record\", \"fields\": [{\"name\": \"data\", \"type\": \"float\"}]}"}' http://localhost:8081/subjects/baz/versions
{"id":1}

% curl http://localhost:8081/subjects/baz/versions/2
{"subject":"baz","version":2,"id":1,"schema":"{\"type\":\"record\",\"name\":\"Foo2\",\"fields\":[{\"name\":\"data\",\"type\":\"float\"}]}"}
```

Is the real bug here that we're supposed to keep track of the next version and if `version=0` set the version to be the next version?

michael-redpanda replied:
It's not being rejected.  Setting to `invalid_schema_version` ensures that the next valid version ID will be calculated

pgellert replied:
Thanks, now I see how this works. Thanks for adding it to the comment as well.

pgellert commented on src/v/pandaproxy/schema_registry/seq_writer.cc:
```diff
@@ -225,7 +225,10 @@ ss::future<std::optional<schema_id>> seq_writer::do_write_subject_version(
           });
 
-    if (!projected.inserted) {
+    if (projected.version == invalid_schema_version) {
+        // Unable to project a version because we are at max
+        throw as_exception(versions_exhausted(schema.schema.sub()));
+    } else if (!projected.inserted) {
```
Should we move this check down into `_store.project_ids()` or even `_store.project_verrsion()` to ensure that it gets logged above at debug level (`"write_subject_version: project_ids failed: {}"`)?

Also, I can see that if the version is exhausted, we do return an `invalid_schema_version`; but I am not sure if `invalid_schema_version` always implies `versions_exhausted`. I think if the schema version already exists then we also get that `projected.version == invalid_schema_version` here.

michael-redpanda replied:
yeah good catch here

michael-redpanda commented:
Force push [2402fc5](https://github.com/redpanda-data/redpanda/commit/2402fc5d637aab2f63d77ae6bf0803a603dc0441):

* Addressed PR comments - fixed issue with version exhaustion check
* Fixed lint issues
* added tests

pgellert reviewed (dismissed):
lgtm

BenPope commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -390,6 +391,12 @@ class store {
 
+        // Once we have hit the maximum version number, we can't continue on
+        if (maxver == std::numeric_limits<schema_version::type>::max()) {
+            vlog(plog.info, "Exhausted available versions for subject {}", sub);
```
The store doesn't usually log, since errors are already logged with more context:
```
INFO  2025-01-23 18:09:47,294 [shard 0:main] pandaproxy - store.h:396 - Exhausted available versions for subject simple-value
WARN  2025-01-23 18:09:47,294 [shard 0:main] pandaproxy - server.cc:132 - Request: POST /subjects/simple-value/versions failed: pandaproxy::schema_registry::exception (Versions exhausted for subject simple-value)
```
You could argue the line number is useful, and I could argue there's only one call site.

michael-redpanda commented:
Force push [821a040](https://github.com/redpanda-data/redpanda/commit/821a0402bea7dd2abbc133192b99b52cc88a0f34):

* Remove unnecessary log line

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24860-v24.1.x-411 remotes/upstream/v24.1.x
git cherry-pick -x c68eeef9eb 821a0402be
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12939503500).
